### PR TITLE
Bundle splitting

### PR DESCRIPTION
**Pull request by janklimo**
#### Changes

1. Update dependencies. `rubocop-performance` is added as a part of Rubocop update.
1. Split JS bundle into chunks.
 
#### Before:
Screen Shot 2019-04-23 at 2 49 39 PM

#### After:
Screen Shot 2019-04-23 at 2 49 49 PM

### CODE REVIEW

**Comment by hiattp** [comment]
@janklimo seems like we should also add the pack builder and/or a README addition to describe how we recommend chunks be included in the head vs body now that they can be split

**Comment by janklimo** [comment]
@hiattp the reason I didn't is that once we use `react-rails`, there will only ever be one pack. We'll split it into chunks for optimal delivery but no chunks will go into the body. I think that's optimal: a) perfect for TL, b) simpler.

**Comment by hiattp** [comment]
@janklimo will that prevent us from "lazy loading" page-specific scripts? Part of the motivation (if not the primary motivation) for putting scripts in the body was to avoid loading all components for all pages. I anticipate having a good amount of page-specific js code as the app grows, ideally we only load that on an as-needed basis

**Comment by janklimo** [comment]
@hiattp that would have been the ideal solution without TL (when we request all assets all the time) but with TL involved, I came to the conclusion that nothing beats this setup. Here's my thinking.

Firstly, savings of not including everything in the header are marginal.

Screen Shot 2019-04-24 at 3 43 34 PM

This shows that vendor chunk represents 84% of our gzipped JS code, the rest being page-specific chunks. Even as our app grows, I don't see this falling below ~70%. The difference between downloading only vendor chunk or all chunks on initial load will be sub-100ms, unnoticeable to the user.

So we have initial page load that is marginally slower, but this is paid back many times over with multiple page visits.

With TL, if the packs are in the head all the time, we never fetch and re-evaluate them after the initial page load. Here's navigating between pages with chunked bundle in the head:


Screen Shot 2019-04-23 at 2 49 39 PM

On the other hand, if we put page-specific code in the body, the browser will fetch and run it. Fetching is instantaneous when the scripts are cached by the browser but evaluation is not.

Screen Shot 2019-04-24 at 4 21 11 PM

Here (current staging) we're delaying DOM loaded event (which other listeners may depend on) by 90ms. We get this delay on every single page visit.

In conclusion, I think loading all JS on page load will not be noticeable to the user but it will make all of the subsequent transitions a bit faster, especially on mobile, where script evaluation takes longer.

**Comment by hiattp** [comment]
Yeah seems like this is the way to go, just get everything up front since loading page-specific stuff incrementally adds too much complexity and/or isn't possible while yielding little benefit.

This will always be the authentication page for many applications (or at least we could ensure that), so we could theoretically require that page not use pack js and lazy load all of the js into the head at that point, so the root page stays fast and you still get all of the heavy JS loading while they user logs in. That seem like a reasonable/possible solution for improved performance if we can guarantee limited js on auth?

**Comment by janklimo** [comment]
Good point about reduced complexity 👍 that's a nice bonus. From what I've seen, `react-rails` also favors convention, so you don't need to register components in your pack before you can use them in the views. Their path is inferred and they're required automatically (so we don't need to modify the pack entry file at all).

Yup we can make the scripts on the login page `async defer` to make it even faster.

**Comment by janklimo** [comment]
@hiattp just noticed bundle splitting commit doesn't seem to have been merged into vmx development

**Comment by hiattp** [comment]
Yeah I haven't merged the latest template code in since the ruby update...feel free to merge it in otherwise I was going to wait until another round of lib updates since this isn't critical (and might require some larger refactors of our current pack/js inclusion, which isn't worth the effort atm)
